src/Raven.Server/Web/Studio/Sharding/Processors/ShardedStudioCollectionsHandlerProcessorForPreviewCollection.cs

```diff
@@ -60,135 +91,89 @@ protected override bool NotModified(out string etag)
     }
 
     public override void Dispose()
     {
+        _combinedReadState.Dispose();
         base.Dispose();
 
         _releaseContext?.Dispose();
         _releaseContext = null;
     }
 
-    private class PreviewCollectionResult
+    public class PreviewCollectionResult
     {
         public List<Document> Results;
         public long TotalResults;
         public List<string> AvailableColumns;
     }
 
-    private readonly struct ShardedCollectionPreviewOperation : IShardedOperation<PreviewCollectionResult>
+    private readonly struct ShardedCollectionPreviewOperation : IShardedStreamableOperation
     {
         private readonly ShardedRequestHandler _handler;
         private readonly ShardedPagingContinuation _token;
-        private readonly JsonOperationContext _context;
 
         public ShardedCollectionPreviewOperation(ShardedRequestHandler handler, ShardedPagingContinuation token, JsonOperationContext context)
         {
             _handler = handler;
             _token = token;
-            _context = context;
         }
 
-        public PreviewCollectionResult Combine(Memory<PreviewCollectionResult> results)
-        {
-            var availableColumns = new HashSet<string>(StringComparer.InvariantCultureIgnoreCase);
-            var total = new PreviewCollectionResult
-            {
-                Results = new List<Document>()
-            };
+        public RavenCommand<StreamResult> CreateCommandForShard(int shard) =>
+            new ShardedCollectionPreviewCommand(_handler, _token.Pages[shard].Start, _token.PageSize);
 
-            var span = results.Span;
-            var totalDocuments = 0L;
+        private class ShardedCommandAsStream : ShardedBaseCommand<StreamResult>
+        {
+            public override bool IsReadRequest => true;
 
-            for (int i = 0; i < span.Length; i++)
+            public ShardedCommandAsStream(ShardedRequestHandler handler, Headers headers) : base(handler, headers, content: null)
             {
-                availableColumns.AddRange(span[i].AvailableColumns);
-                total.TotalResults += span[i].TotalResults;
             }
 
-            for (int i = 0; i < span.Length * _token.PageSize; i++)
+            public override async Task<ResponseDisposeHandling> ProcessResponse(JsonOperationContext context, HttpCache cache, HttpResponseMessage response, string url)
             {
-                var takeFromShard = i % _token.Pages.Length;
+                var responseStream = await response.Content.ReadAsStreamAsync().ConfigureAwait(false);
 
-                var shardDocs = span[takeFromShard].Results;
-                if (shardDocs.Count == 0)
-                    continue;
-
-                if (totalDocuments++ >= _token.PageSize)
-                    return total;
+                Result = new StreamResult
+                {
+                    Response = response,
+                    Stream = new StreamWithTimeout(responseStream)
```

Comment:
do we want to wrap it? there is a read timeout there and I'm assuming that some of the streams might not be read for a long time?

Reply:
This is not something that we should do across the board, but I do expect the studio to show me the results after 2 min..